fast.py

```diff
@@ -1,58 +1,130 @@
-            g = self.triple_re.search(triple)
-            if not g:
-                continue
-            identifier, name = g.groups()
-            self[identifier] = name
-
     @classmethod
     def from_data_directory(cls, data_directory):
+        """Load names from a directory that either contains a bunch of
+        files in N-Triples format or a single consolidated CSV file.
+        """
         my_directory = os.path.join(data_directory, cls.SUBDIR)
-        names = cls()
         consolidated_file = os.path.join(my_directory, "consolidated.csv.gz")
         a = time.time()
         if os.path.exists(consolidated_file):
-            logging.info("Reading cached %s names from %s",
-                cls.SUBDIR, consolidated_file)
-            input_file = gzip.open(consolidated_file)
-            reader = csv.reader(input_file)
-            for k, v in reader:
-                names[k] = v
+            # A consolidated file has already been created. Load it --
+            # it's quick.
+            names = cls.from_consolidated_file(consolidated_file)
         else:
-            for i in os.listdir(my_directory):
-                if not i.endswith(".nt.gz") and not i.endswith(".nt"):
-                    continue
+            # We have to go through a bunch of N-Triples files.
+            names = cls()
+            for i in sorted(os.listdir(my_directory)):
                 path = os.path.join(my_directory, i)
                 logging.info("Loading %s names from %s", cls.SUBDIR, path)
-                names.load_filehandle(gzip.open(path))
+                names.load_triples_file(path)
                 logging.info(
-                    "There are now %d %s names.", cls.SUBDIR, len(names))
+                    "There are now %d %s names.", len(names), cls.SUBDIR
+                )
 
-            output = gzip.open(consolidated_file,"w")
-            writer = csv.writer(output)
-            for k,v in names.items():
-                writer.writerow([k, v])
-            output.close()
+            # Now that we've done that, write out a consolidated file
+            # so next time will go more quickly.
+            names.write_consolidated_file(consolidated_file)
         b = time.time()
-        logging.info("Done loading %s names in %.1f sec", cls.SUBDIR, b-a)
+        logging.info(
+            "Loaded %d %s names in %.1f sec", len(names), cls.SUBDIR, (b-a)
+        )
+        return names
+
+    def load_triples_file(self, path):
+        """Load classifications from an N-Triples file."""
+        if path.endswith(".nt.gz"):
+            # This is a single GZipped N-Triples file.
+            self.load_triples_filehandle(gzip.open(path, 'rb'))
+        elif path.endswith(".nt.zip"):
+            # This is a ZIP file containing one or more (probably just
+            # one) N-Triples files. Load each one the zip.
+            for fh in self.triples_filehandles_from_zip(path):
+                self.load_triples_filehandle(fh)
```

Comment:
Previously, we handled .nt.gz and .nt files, but ignored .nt.zip files. But .nt.zip is what we actually get from the LCSH dump -- a bunch of ZIP files, each with an .nt file buried deep within its directory structure. I introduced triples_filehandles_from_zip to deal with this.